File: ios/Classes/SDK/AuthenticationResultMapping.swift
import Foundation
import Gateway

/// Converts the iOS SDK authentication response into the contract.
///
/// Payment outcomes (declines, cancellation) become result messages; integration and technical
/// failures become channel errors. Authentication requests carry no card data, so SDK error
/// texts are kept as native details to help diagnose gateway configuration problems.
func toAuthenticationResult(
  _ response: AuthenticationResponse,
  authenticationTransactionId: String
) -> Result<AuthenticationResultMessage, Error> {
  func outcome(_ outcome: AuthenticationOutcomeMessage) -> Result<AuthenticationResultMessage, Error> {
    return .success(
      AuthenticationResultMessage(
        outcome: outcome,
        authenticationPerformed: response.authenticationPerformed,
        challengePerformed: response.challengePerformed,
        authenticationTransactionId: authenticationTransactionId,
        sdkTransactionId: response.sdkTransactionId,
        threeDS2TransactionStatus: response.threeDS2TransactionStatus))
  }

  if response.recommendation == .proceed {
    return outcome(.proceed)
  }

  guard let error = response.error else {
    return outcome(.doNotProceed)
  }

  guard let authenticationError = error as? AuthenticationError else {
    // Network and gateway failures from the authentication requests.
    return .failure(gatewayRequestBridgeError(error))
  }

  switch authenticationError {
  case .challengeCancelledByUser:
    return outcome(.cancelledByUser)
  case .challengeTimedOut:
    return outcome(.challengeTimedOut)
  case .recommendation_ResubmitWithAlternativePaymentDetails:
    return outcome(.resubmitWithAlternativePaymentDetails)
  case .recommendation_AbandonOrder:
    return outcome(.abandonOrder)
  case .recommendation_DoNotProceed:
    return outcome(.doNotProceed)
  case .recommendation_Unknown:
    return outcome(.unknownRecommendation)
  case .notInitialized:
    return .failure(
      gatewayBridgeError(
        code: errorCodeNotInitialized,
        message: "The Gateway SDK is not initialized."))
  case .missingParameter(let parameter):
    return .failure(
      gatewayBridgeError(
        code: errorCodeMissingSessionParameter,
        message: "The session is missing fields required for authentication.",
        nativeDetails: parameter))
  case .invalidChallengeCompletionURL(let detail):
    return .failure(
      gatewayBridgeError(
        code: errorCodeInvalidChallengeCompletionUrl,
        message: "The 3DS challenge completion URL is invalid.",
        nativeDetails: detail))
  case .other(let detail):
    return .failure(
      gatewayBridgeError(
        code: errorCodeUnknown,
        message: "Payer authentication failed.",
        nativeDetails: "AuthenticationError.other: \(detail)"))
  @unknown default:
    // A case added by a newer SDK version.
    return .failure(
      gatewayBridgeError(
        code: errorCodeUnknown,
        message: "Payer authentication failed.",
        nativeDetails: "AuthenticationError: \(authenticationError)"))
  }
}

File: android/src/main/kotlin/com/example/nbe_payment_flutter_plugin/sdk/AuthenticationResultMapping.kt
package com.example.nbe_payment_flutter_plugin.sdk

import com.example.nbe_payment_flutter_plugin.bridge.gatewayBridgeError
import com.example.nbe_payment_flutter_plugin.bridge.sanitizedSdkDetail
import com.example.nbe_payment_flutter_plugin.generated.AuthenticationOutcomeMessage
import com.example.nbe_payment_flutter_plugin.generated.AuthenticationResultMessage
import com.example.nbe_payment_flutter_plugin.generated.errorCodeInvalidChallengeCompletionUrl
import com.example.nbe_payment_flutter_plugin.generated.errorCodeMissingSessionParameter
import com.example.nbe_payment_flutter_plugin.generated.errorCodeNotInitialized
import com.example.nbe_payment_flutter_plugin.generated.errorCodeUnknown
import com.mastercard.gateway.android.sdk.AuthenticationError
import com.mastercard.gateway.android.sdk.AuthenticationRecommendation
import com.mastercard.gateway.android.sdk.AuthenticationResponse

/**
 * Converts the Android SDK authentication response into the contract.
 *
 * The SDK reports every outcome through `AuthenticationResponse`: payment outcomes (declines,
 * cancellation) become result messages, while integration and technical failures become
 * channel errors.
 *
 * Authentication requests carry no card data (the card is already in the session), so SDK error
 * texts help diagnose gateway configuration problems. They are unbounded third-party text
 * (from the SDK, the 3DS server or the issuer), so they are shortened before being sent.
 */
internal fun toAuthenticationResult(
    response: AuthenticationResponse,
    authenticationTransactionId: String,
): Result<AuthenticationResultMessage> {
    fun outcome(outcome: AuthenticationOutcomeMessage) = Result.success(
        AuthenticationResultMessage(
            outcome = outcome,
            authenticationPerformed = response.authenticationPerformed,
            challengePerformed = response.challengePerformed,
            authenticationTransactionId = authenticationTransactionId,
            // Not reported by the Android SDK.
            sdkTransactionId = null,
            threeDS2TransactionStatus = null,
        ),
    )

    if (response.recommendation == AuthenticationRecommendation.PROCEED) {
        return outcome(AuthenticationOutcomeMessage.PROCEED)
    }

    return when (val error = response.error) {
        null -> outcome(AuthenticationOutcomeMessage.DO_NOT_PROCEED)

        is AuthenticationError.ChallengeCancelledByUser ->
            outcome(AuthenticationOutcomeMessage.CANCELLED_BY_USER)
        is AuthenticationError.ChallengeTimedOut ->
            outcome(AuthenticationOutcomeMessage.CHALLENGE_TIMED_OUT)
        is AuthenticationError.RecommendationResubmitWithAlternativePaymentDetails ->
            outcome(AuthenticationOutcomeMessage.RESUBMIT_WITH_ALTERNATIVE_PAYMENT_DETAILS)
        is AuthenticationError.RecommendationAbandonOrder ->
            outcome(AuthenticationOutcomeMessage.ABANDON_ORDER)
        is AuthenticationError.RecommendationDoNotProceed ->
            outcome(AuthenticationOutcomeMessage.DO_NOT_PROCEED)
        is AuthenticationError.RecommendationUnknown ->
            outcome(AuthenticationOutcomeMessage.UNKNOWN_RECOMMENDATION)

        is AuthenticationError.NotInitialized -> Result.failure(
            gatewayBridgeError(
                code = errorCodeNotInitialized,
                message = "The Gateway SDK is not initialized.",
                nativeDetails = sanitizedSdkDetail(error.error),
            ),
        )
        is AuthenticationError.MissingParameter -> Result.failure(
            gatewayBridgeError(
                code = errorCodeMissingSessionParameter,
                message = "The session is missing fields required for authentication.",
                nativeDetails = sanitizedSdkDetail(error.error),
            ),
        )
        is AuthenticationError.InvalidChallengeCompletionURL -> Result.failure(
            gatewayBridgeError(
                code = errorCodeInvalidChallengeCompletionUrl,
                message = "The 3DS challenge completion URL is invalid.",
                // Everything after "?" is dropped: a challenge URL's query can carry tokens.
                nativeDetails = sanitizedSdkDetail(error.error.substringBefore('?')),
            ),
        )
        is AuthenticationError.Other -> Result.failure(
            gatewayBridgeError(
                code = errorCodeUnknown,
                message = "Payer authentication failed.",
                nativeDetails = sanitizedSdkDetail("AuthenticationError.Other: ${error.error}"),
            ),
        )

        // Network and gateway failures from the authentication requests are surfaced raw by
        // the SDK, exactly like update-session failures.
        else -> Result.failure(gatewayRequestBridgeError(error))
    }
}
